apps/examples/mediarecorder/mediarecorder_main.cpp

```diff
@@ -164,61 +199,59 @@ class MediaRecorderTest : public MediaRecorderObserverInterface, public enable_s
 					}
 				}
 				break;
-				
+
 			}
 		}
 	}
 
 	void play_data()
 	{
-		int fd;
-		char *buffer;
-		int num_read;
-		unsigned int size;
+		cout << "playback " << filePath << endl;
 
-		p_out = pcm_open(0, 0, PCM_OUT, &pcmConfig);
+		auto source = std::move(unique_ptr<FileInputDataSource>(new FileInputDataSource(filePath)));
+		source->setSampleRate(16000);
+		source->setChannels(2);
+		source->setPcmFormat(AUDIO_FORMAT_TYPE_S16_LE);
 
-		if (pcm_get_file_descriptor(p_out) < 0) {
-			printf("pcm open fail\n%s\n", pcm_get_error(p_out));
+		MediaPlayer mp;
+		if (mp.create() == PLAYER_ERROR) {
+			cout << "Mediaplayer::create failed" << endl;
 			return;
 		}
 
-		size = pcm_frames_to_bytes(p_out, pcm_get_buffer_size(p_out));
-		buffer = (char *)malloc(size);
+		do {
```

Comment:
we don't need loop here,
prepare -> start
and when it reaches end of file it will stop automatically and onPlaybackFinishged will be called
Then we can destroy player there.

Reply:
I use `do { break; } while (0);` which is not a real loop, just break out the playing sequences and go to the last `destroy` operation.

Reply:
I mean F/W has it own thread to handle, we don't have to play inside of loop, that's why we added thread & callback interface.
This application is one of example for user, we'd better follow generic usage.

Reply:
Do you mean we don't need to wait playing finished here (in play_data()), 
just do unprepare&destroy in onPlaybackFinishged() ?
play_data() only call mediaplayer to start playback and then exit at once, right?